src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Ascii.CaseConversion.cs

```diff
@@ -0,0 +1,507 @@
+
+        Drain32:
+
+            // Attempt to process blocks of 32 input bits.
+
+            if ((elementCount - i) >= (nuint)(4 / sizeof(TFrom)))
+            {
+                uint nextBlockAsUInt32 = Unsafe.ReadUnaligned<uint>(&pSrc[i]);
+                if (SourceIsAscii)
+                {
+                    if (!Utf8Utility.AllBytesInUInt32AreAscii(nextBlockAsUInt32))
+                    {
+                        goto DrainRemaining;
+                    }
+                    nextBlockAsUInt32 = (ConversionIsToUpper)
+                        ? Utf8Utility.ConvertAllAsciiBytesInUInt32ToUppercase(nextBlockAsUInt32)
+                        : Utf8Utility.ConvertAllAsciiBytesInUInt32ToLowercase(nextBlockAsUInt32);
+                }
+                else
+                {
+                    if (!Utf16Utility.AllCharsInUInt32AreAscii(nextBlockAsUInt32))
+                    {
+                        goto DrainRemaining;
+                    }
+                    nextBlockAsUInt32 = (ConversionIsToUpper)
+                        ? Utf16Utility.ConvertAllAsciiCharsInUInt32ToUppercase(nextBlockAsUInt32)
+                        : Utf16Utility.ConvertAllAsciiCharsInUInt32ToLowercase(nextBlockAsUInt32);
+                }
+
+                if (ConversionIsWidthPreserving)
+                {
+                    Unsafe.WriteUnaligned<uint>(&pDest[i], nextBlockAsUInt32);
+                }
+                else
+                {
+                    Debug.Assert(Vector128.IsHardwareAccelerated);
+
+                    Vector128<uint> blockAsVectorOfUInt32 = Vector128.CreateScalarUnsafe(nextBlockAsUInt32);
+                    if (ConversionIsWidening)
+                    {
+                        Vector128<ulong> widenedBlock = Vector128.WidenLower(blockAsVectorOfUInt32.AsByte()).AsUInt64();
+                        Unsafe.WriteUnaligned<ulong>(&pDest[i], widenedBlock.ToScalar());
+                    }
+                    else
+                    {
+                        Vector128<ushort> blockAsVectorOfUInt16 = blockAsVectorOfUInt32.AsUInt16();
+                        Vector128<ushort> narrowedBlock = Vector128.Narrow(blockAsVectorOfUInt16, blockAsVectorOfUInt16).AsUInt16();
+                        Unsafe.WriteUnaligned<ushort>(&pDest[i], narrowedBlock.ToScalar());
+                    }
+                }
+
+                i += (nuint)(4 / sizeof(TFrom));
+
+                // If vectorization is not accelerated or we're on 32-bit,
+                // turn this into a while loop.
+
+                if (IntPtr.Size < 8 || !Vector128.IsHardwareAccelerated)
+                {
+                    goto Drain32;
+                }
+            }
+
+        DrainRemaining:
+
+            // Process single elements at a time.
+
+            for (; i < elementCount; i++)
+            {
+                uint element = uint.CreateTruncating(pSrc[i]);
+                if (!UnicodeUtility.IsAsciiCodePoint(element)) { break; }
```

Comment:
Nit: let's avoid single-line if blocks with bodies like this, especially in the context of all the other blocks being multi-line